pywr-schema/src/timeseries/align_and_resample.rs

```diff
@@ -0,0 +1,245 @@
+use chrono::TimeDelta;
+use polars::{prelude::*, series::ops::NullBehavior};
+use pywr_core::models::ModelDomain;
+use std::{cmp::Ordering, ops::Deref};
+
+use crate::timeseries::TimeseriesError;
+
+pub fn align_and_resample(
+    name: &str,
+    df: DataFrame,
+    time_col: &str,
+    domain: &ModelDomain,
+) -> Result<DataFrame, TimeseriesError> {
+    // Ensure type of time column is datetime and that it is sorted
+    let df = df
+        .clone()
+        .lazy()
+        .with_columns([col(time_col).cast(DataType::Datetime(TimeUnit::Nanoseconds, None))])
+        .collect()?
+        .sort([time_col], false, true)?;
+
+    // Ensure that df start aligns with models start for any resampling
+    let df = slice_start(df, time_col, domain)?;
+
+    // Get the durations of the time column
+    let durations = df
+        .clone()
+        .lazy()
+        .select([col(time_col).diff(1, NullBehavior::Drop).unique().alias("duration")])
+        .collect()?;
+    let durations = durations.column("duration")?.duration()?.deref();
+
+    if durations.len() > 1 {
+        // Non-uniform timestep are not yet supported
+        todo!();
+    }
+
+    let timeseries_duration = match durations.get(0) {
+        Some(duration) => duration,
+        None => return Err(TimeseriesError::TimeseriesDurationNotFound(name.to_string())),
+    };
+
+    let model_duration = domain
+        .time()
+        .step_duration()
+        .whole_nanoseconds()
+        .expect("Nano seconds could not be extracted from model step duration");
+
+    let df = match model_duration.cmp(&timeseries_duration) {
+        Ordering::Greater => {
+            // Downsample
+            df.clone()
+                .lazy()
+                .group_by_dynamic(
+                    col(time_col),
+                    [],
+                    DynamicGroupOptions {
+                        every: Duration::new(model_duration),
+                        period: Duration::new(model_duration),
+                        offset: Duration::new(0),
+                        start_by: StartBy::DataPoint,
+                        ..Default::default()
+                    },
+                )
+                .agg([col("*").exclude([time_col]).mean()])
+                .collect()?
+        }
+        Ordering::Less => {
+            // Upsample
+            // TODO: this does not extend the dataframe beyond its original end date. Should it do when using a forward fill strategy?
```

Comment:
Yes, I think so.

Reply:
I might do this one in a new PR if that's ok as its going to require a bit of testing